design/spark-co-exist-with-object-storages.md

```diff
@@ -20,122 +18,178 @@ Make the solution usable by non-lakeFS users by:
+routerfs.mapping.${mappingIdx}.${toFsScheme}.with='${toFsScheme}://another-bucket/prefix/*'
+```
+
+Where `mappingIdx` is an unbounded running index initialized for each `toFsScheme`, and `toFsScheme` is a URI scheme that
+uses the `fs.toFsScheme.impl` property to point to the file system that handles the interaction with the underlying storage.
+For example, the following mapping configurations
+```properties
+routerfs.mapping.1.lakefs.replace='^s3a://bucket/prefix'
+routerfs.mapping.1.lakefs.with='lakefs://example-repo/dev/prefix'
+```
+together with 
+```properties
+fs.lakefs.impl=S3AFileSystem
+```
+make `RouterFileSystem` translate `s3a://bucket/prefix/foo.parquet` into `lakefs://example-repo/dev/prefix/foo.parquet`, 
+and later use `S3AFileSystem` to interact with the underlying object storage which is S3 in this case. This example uses
+[S3 gateway](../docs/integrations/spark.md#access-lakefs-using-the-s3a-gateway) as the Spark-lakeFS integration method.
+
+##### Multiple mapping configurations   
+
+Each `toFsScheme` can have any number of mapping configurations. E.g., below are two mapping configurations for `toFsScheme=lakefs`.
+```properties
+routerfs.mapping.1.lakefs.replace='^s3a://bucket/prefix'
+routerfs.mapping.1.lakefs.with='lakefs://example-repo/dev/prefix'
+routerfs.mapping.2.lakefs.replace='^s3a://another-bucket'
+routerfs.mapping.2.lakefs.with='lakefs://another-repo/main'
+```
+Mapping configurations applied in order, therefore in case of a conflict in mapping configurations the prior configuration 
+applies. 
+
+##### Multiple `toFsScheme`s (and multiple mapping configuration groups)
+
+With `RouterFileSystem`, Spark users can define any number of `toFsScheme`s. Each forms its own mapping configuration group,
+and allows applying different set of Spark/Hadoop configurations. e.g. credentials, s3 endpoint, etc. Users would typically
+define new `toFsScheme` while trying to migrate a collection from one storage space to another without changing their Spark code.
+
+The example below demonstrates how routerFS mapping configurations and some Hadoop configurations will look like for
+Spark application that accesses s3, MinIO, and lakeFS, but is using s3a as its sole URI scheme. 
+```properties
+# Mapping configurations
+routerfs.mapping.1.lakefs.replace='^s3a://bucket/prefix'
+routerfs.mapping.1.lakefs.with='lakefs://example-repo/dev/prefix'
+routerfs.mapping.1.s3a1.replace='^s3a://bucket'
+routerfs.mapping.1.s3a1.with='s3a1://another-bucket'
+routerfs.mapping.1.minio.replace='^s3a://minio-bucket'
+routerfs.mapping.1.minio.with='minio://another-minio-bucket'
+
+# File System configurations
+fs.s3a.impl=RouterFileSystem
+fs.lakefs.impl=S3AFileSystem
+fs.s3a1.impl=S3AFileSystem
+fs.minio.impl=S3AFileSystem
+```
+
+##### Default mapping configuration 
+
+`RouterFileSystem` requires a default mapping configuration in case that none of the mappings matches a URI. 
```

Comment:
Is this even required?  I mean it's obviously a best practice, but I reckon it's OK to document what happens if you don't do this.  For instance, suppose it failed on error if no mapping matched.  Why is that a bad choice?  S3A with default settings does the same thing if I give it a bad URL: my access is denied or I get some other error message, and I know what to do.

Reply:
Without a default mapping, how would routerFS find an underlying filesystem that will invoke the file system operation? 
For example, given the following configurations 
```properties
routerfs.mapping.1.lakefs.replace='^s3a://bucket/prefix'
routerfs.mapping.1.lakefs.with='lakefs://example-repo/dev/prefix'

# File System configurations
fs.s3a.impl=RouterFileSystem
fs.lakefs.impl=S3AFileSystem
```
and the path `"s3a://another-bucket/prefix'"`
What underlying fs RouterFS should use to invoke a `fs.open("s3a://another-bucket/prefix")`?

Please let me know if I'm missing something!